src/main/java/com/iabtcf/v2/BitVectorTCModelV2.java

```diff
@@ -19,296 +21,364 @@
     }
 
     @Override
     public boolean isServiceSpecific() {
-        return this.isServiceSpecific;
+        if (cache.add(CORE_IS_SERVICE_SPECIFIC)) {
+            isServiceSpecific = bbv.readBits1(CORE_IS_SERVICE_SPECIFIC);
+        }
+        return isServiceSpecific;
     }
 
     @Override
     public boolean useNonStandardStacks() {
-        return this.useNonStandardStacks;
+        if (cache.add(CORE_USE_NON_STANDARD_STOCKS)) {
+            useNonStandardStacks = bbv.readBits1(CORE_USE_NON_STANDARD_STOCKS);
+        }
+        return useNonStandardStacks;
     }
 
     @Override
     public Set<Integer> specialFeatureOptIns() {
-        return this.specialFeaturesOptInts;
+        if (cache.add(CORE_SPECIAL_FEATURE_OPT_INS)) {
+            specialFeaturesOptInts = fillSet(
+                    CORE_SPECIAL_FEATURE_OPT_INS.getOffset(bbv),
+                    CORE_SPECIAL_FEATURE_OPT_INS.getLength(bbv),
+                    bbv);
+        }
+        return specialFeaturesOptInts;
     }
 
     @Override
     public Set<Integer> purposesConsent() {
-        return this.purposesConsent;
+        if (cache.add(CORE_PURPOSES_CONSENT)) {
+            purposesConsent = fillSet(
+                    CORE_PURPOSES_CONSENT.getOffset(bbv),
+                    CORE_PURPOSES_CONSENT.getLength(bbv),
+                    bbv);
+        }
+        return purposesConsent;
     }
 
     @Override
     public Set<Integer> purposesLITransparency() {
-        return this.purposesLITransparency;
+        if (cache.add(CORE_PURPOSES_LI_TRANSPARENCY)) {
+            purposesLITransparency = fillSet(
+                    CORE_PURPOSES_LI_TRANSPARENCY.getOffset(bbv),
+                    CORE_PURPOSES_LI_TRANSPARENCY.getLength(bbv),
+                    bbv);
+        }
+        return purposesLITransparency;
     }
 
     @Override
     public boolean isPurposeOneTreatment() {
-        return this.isPurposeOneTreatment;
+        if (cache.add(CORE_PURPOSE_ONE_TREATMENT)) {
+            isPurposeOneTreatment = bbv.readBits1(CORE_PURPOSE_ONE_TREATMENT);
+        }
+        return isPurposeOneTreatment;
     }
 
     @Override
     public String publisherCountryCode() {
-        return this.publisherCountryCode;
+        if (cache.add(CORE_PUBLISHER_CC)) {
+            publisherCountryCode = ByteBitVectorUtils.readStr2(bbv, CORE_PUBLISHER_CC);
+        }
+        return publisherCountryCode;
     }
 
     @Override
     public Set<Integer> vendorConsents() {
-        return this.vendorConsents;
+        if (cache.add(CORE_VENDOR_BITRANGE_FIELD)) {
+            vendorConsents = new TreeSet<>();
+            fetchSet(vendorConsents, CORE_VENDOR_MAX_VENDOR_ID.getOffset(bbv), bbv);
+        }
+        return vendorConsents;
     }
 
     @Override
     public Set<Integer> vendorLegitimateInterests() {
-        return this.vendorLegitimateInterests;
+        if (cache.add(CORE_VENDOR_LI_BITRANGE_FIELD)) {
+            vendorLegitimateInterests = new TreeSet<>();
```

Comment:
would like to see these use HashSet instead of TreeSet. Can wait for later commit but we are already changing all of the logic here, wouldn't be much more to include now.

Reply:
yeah, i agree. let's revisit this in another pr. good call out though.